Add timestamp and scheduled finalisation time to InvalidExit and UnchallengedExit events

okalouti opened the pull request:
:clipboard: Closes #1481 

## Overview

This adds exit timestamp and scheduled finalisation time to standard exit structs persisted in the DB - allowing the latter to be reported in `invalid_exit` and `unchallenged_exit` events. 

```json
"details": {
          "amount": 12000000000000000000,
          "currency": "0x0000000000000000000000000000000000000000",
          "eth_height": 325,
          "name": "invalid_exit",
          "owner": "0xc29320f6a614edd1ffc8c7f8a7c8dd703761a68a",
          "root_chain_txhash": "0xee0d9e06dedf792d25c3d168fe1073c42039b49ddeffd76fc57e31201df8887b",
          "scheduled_finalization_time": 1588144725,
          "utxo_pos": 1000000000
        },
        "event": "invalid_exit"
      }
```

## Changes


-  Add logic in the `ExitProcessor` (`handle_call({:new_exits, exits}`) to fetch timestamp and calculate scheduled finalisation time after making the requisite contract calls, then adding them to the exit event.
-  Make requisite changes in `ExitInfo` (including adding a helper to calculate SFT)

## Testing

- Updated relevant tests to have dummy `timestamp` `scheduled_finalisation_time` and updated tests in `exit_info_test.exs`

Specs: 

- Updated Swagger specs to reflect these changes along with addition of `root_chain_txhash` and `spending_txhash` 

Note: adding an exit timestamp is a complement to this PR changing SLA margin from blocks to seconds. This will enable a straightforward segmentation of `invalid` vs. `unchallenged` exits.

InoMurko commented on apps/omg_watcher/lib/omg_watcher/exit_processor.ex:
```diff
@@ -652,4 +657,17 @@ defmodule OMG.Watcher.ExitProcessor do
+    {:utxo_position, blknum, _, _} = Utxo.Position.decode!(utxo_pos_enc)
+    {_block_hash, utxo_creation_block_timestamp} = RootChain.blocks(blknum)
+    {:ok, exit_block_timestamp} = Eth.get_block_timestamp_by_number(eth_height)
+    min_exit_period = Configuration.min_exit_period_seconds()
```
if you need min exit period it needs to be in the process state, you cannot access configuration from here

InoMurko commented on apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:
```diff
@@ -34,8 +35,12 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
+    :scheduled_finalization_time
   ]
+
+  @child_block_interval Eth.Configuration.child_block_interval()
```
this can't be hardcoded at compile time

InoMurko commented on apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:
```diff
@@ -137,4 +147,21 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
+      do: {:ok, max(exit_timestamp + min_exit_period, utxo_creation_timestamp + min_exit_period)}
+
+  def calculate_sft(_blknum, exit_timestamp, utxo_creation_timestamp, min_exit_period),
+    do: {:ok, max(exit_timestamp + min_exit_period, utxo_creation_timestamp + 2 * min_exit_period)}
```
these are multiline functions, why squash them into a single line function defintion? use `do...end`

okalouti replied:
Refactored in dd9ac6b.

okalouti commented:
Hey @InoMurko. Thanks for the early feedback. I've added `child_block_interval` and `min_exit_period` in the ExitProcessor state and refactored accordingly. 

What remains in this PR is to add a `timestamp` to the persisted block struct in order to complement @souradeep-das's PR that switches `sla_margin` from blocks to seconds. 

(Having an exit timestamp will facilitate the segmentation of `invalid` vs. `unchallenged` exits, see [here](https://github.com/omisego/elixir-omg/pull/1491/files#diff-778619489606bb4d5badbff08b7ba0aeR87))

souradeep-das commented:
`a6f6b18` timestamp to exit struct LGTM

unnawut commented on apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:
```diff
@@ -137,4 +149,31 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
+  @doc """
+  Based on the block number determines whether UTXO was created by a deposit.
+  """
+  defguard is_deposit(blknum, child_block_interval) when rem(blknum, child_block_interval) != 0
```
I think this can be a private function since we don't use it in pattern matching and it's only used locally in this module.

okalouti replied:
b186e8a

unnawut commented on apps/omg_watcher_rpc/priv/swagger/security_critical_api_specs.yaml:
```diff
@@ -391,6 +391,12 @@ paths:
                                           type: integer
+                                        root_chain_txhash:
+                                          type: string
+                                        spending_txhash:
```
Not sure are we emitting this `spending_txhash` yet?

okalouti replied:
Added in anticipation of Med's recently merged PR: https://github.com/omisego/elixir-omg/pull/1492

okalouti commented on apps/omg_watcher/lib/omg_watcher/exit_processor/core.ex:
```diff
@@ -101,19 +113,31 @@ defmodule OMG.Watcher.ExitProcessor.Core do
+        db_exits,
+        db_in_flight_exits,
+        db_competitors,
+        min_exit_period_seconds \\ @default_min_exit_period_seconds,
```
Agreed. Had not thought about that distinction. Had to update some tests using `Core.init`. 
8387512

okalouti commented on apps/omg_watcher/lib/omg_watcher/exit_processor/exit_info.ex:
```diff
@@ -34,8 +34,11 @@ defmodule OMG.Watcher.ExitProcessor.ExitInfo do
-    :root_chain_txhash
+    :root_chain_txhash,
+    :scheduled_finalization_time,
+    :timestamp
```
b186e8a

okalouti commented on apps/omg_watcher/lib/omg_watcher/exit_processor.ex:
```diff
@@ -290,12 +302,21 @@ defmodule OMG.Watcher.ExitProcessor do
+  def handle_call(
+        {:new_exits, exits},
+        _from,
+        %{min_exit_period_seconds: min_exit_period_seconds, child_block_interval: child_block_interval} = state
```
b186e8a
